clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -205,6 +205,26 @@ public void wakeup() {
             networkClientDelegate.wakeup();
     }
 
+    /**
+     * Returns the delay for which the application thread can safely wait before it should be responsive
+     * to results from the request managers. For example, the subscription state can change when heartbeats
+     * are sent, so blocking for longer than the heartbeat interval might mean the application thread is not
+     * responsive to changes.
+     *
+     * @return The maximum delay in milliseconds
+     */
+    public long maximumTimeToWait() {
+        final long currentTimeMs = time.milliseconds();
+        if (requestManagers == null) {
+            return MAX_POLL_TIMEOUT_MS;
+        }
+        return requestManagers.entries().stream()
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(rm -> rm.maximumTimeToWait(currentTimeMs))
+                .reduce(Long.MAX_VALUE, Math::min);
+    }
+
```

Comment:
This will invoke the request managers directly from the application thread, right?

If so, I'm a little concerned by this approach because the implementation of `maximumTimeToWait` in each `RequestManager` will be executed by the application thread. In the implementations elsewhere in this PR we're reading and writing state in the request managers that is only intended to be read or written by the network I/O thread.

We need to be careful here 🤔

Reply:
Yes, we do. In the previous commit, I inadvertently ended up with a ConcurrentModificationException for this very reason. This is why I scaled back the change to be only accessed the timer.

Reply:
I'm wondering how "fresh" the "maximum time to wait" value needs to be...

Spitballing here, but... say we introduce a new instance variable in `ConsumerNetworkThread` that stores a cached value named, say, `notTotallyFreshMaxWaitCached`? We could then invoke `maximumTimeToWait()` from inside `runOnce()` and update `notTotallyFreshMaxWaitCached`, like so:

```java
    private volatile long notTotallyFreshMaxWaitCached = MAX_POLL_TIMEOUT_MS;

    void runOnce() {
        // If there are errors processing any events, the error will be thrown immediately. This will have
        // the effect of closing the background thread.
        applicationEventProcessor.process();

        final long currentTimeMs = time.milliseconds();
        final long pollWaitTimeMs = requestManagers.entries().stream()
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(rm -> rm.poll(currentTimeMs))
                .map(networkClientDelegate::addAll)
                .reduce(MAX_POLL_TIMEOUT_MS, Math::min);
        networkClientDelegate.poll(pollWaitTimeMs, currentTimeMs);

        notTotallyFreshMaxWaitCached = requestManagers.entries().stream()
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(rm -> rm.maximumTimeToWait(currentTimeMs))
                .reduce(Long.MAX_VALUE, Math::min);
    }

    public long maximumTimeToWait() {
        if (requestManagers == null) {
            return MAX_POLL_TIMEOUT_MS;
        }
        return notTotallyFreshMaxWaitCached;
    }
```

(Or something like that.)

This way we still only touch the `RequestManager`s inside `runOnce()`, so we don't open overselves up to subtle concurrency issues.

Thoughts?

Reply:
Something like this would work I think. The aim after all is to keep things responsive rather than an accurate regular tick.

Reply:
I'll take this concept and work on it in a follow-on PR.